### ISSUE TITLE
Make releases available via Ubuntu PPA

### ISSUE THREAD

**Issue by soredake** [dosbox-staging/dosbox-staging#418]
Would be nice to have one.

**Comment by kcgen** [comment]
As a fellow Ubuntu (mint) user, couldn't agree more :-)

Probably want separate `-stable` / `-release` vs `-latest` / `-nightly` channels too. 

_(I made an attempt at this using a handful of howto's and tutorials a couple months back. Eventually threw in the towel after frustration got the best of me.. plus feels like the whole thing is laden with unnecessary or deliberately convoluted steps; :exploding_head:)._ 

Hoping someone with PPA-packaging experience can offer a hand, in any capacity.  Helpful tips,  Q&A type help, setting up your own staging PPA as a proof-of-concept, or straight-up package the binaries our from the release and CI pages -- all help is good help.

@dreamer and I both have signed launchpad accounts, so can put those to use when the time comes.

**Comment by FeignInt** [comment]
I don't know about PPA, but I have thrown up a POC apt-repo

https://github.com/FeignInt/dosbox-staging-apt-repo-POC

Very recently I added to the debian/rules  (make file) to relax the dependencies, I rambled on about version numbers in the wiki ( linked in the README.md ) 

My hope is that the deb will work on any distro with a passing resemblance to Debian

I have not tested it on anything other than Debian SId, I would be interested to know how far the likes of Mint gets.

Note: in the README.md I included a deb-src line
It is optional
but for kicks
```
apt -s source dosbox-staging
```
-s is simulate, it won't actually download anything.

I hope it helps with packaging debs

**Comment by kcgen** [comment]
So far so good!

### Added key and repo, then updated:

``` text
Get:1 https://raw.githubusercontent.com/feignint/dosbox-staging-apt-repo-POC/master/generic unstable InRelease [2,336 B]
Hit:2 http://packages.microsoft.com/repos/vscode stable InRelease                                                                             
Hit:3 http://mirrors.edge.kernel.org/ubuntu focal InRelease                                                
Get:4 http://mirrors.edge.kernel.org/ubuntu focal-updates InRelease [107 kB]
Get:5 http://mirrors.edge.kernel.org/ubuntu focal-security InRelease [107 kB]              
Get:6 http://mirrors.edge.kernel.org/ubuntu focal-backports InRelease [98.3 kB]                                                                   
Get:7 https://raw.githubusercontent.com/feignint/dosbox-staging-apt-repo-POC/master/generic unstable/otherosfs Sources [744 B]                     
Hit:8 http://archive.canonical.com/ubuntu focal InRelease                                                                                                                    
Get:9 https://raw.githubusercontent.com/feignint/dosbox-staging-apt-repo-POC/master/generic unstable/otherosfs amd64 Packages [1,425 B]    
...
```

### apt show dosbox-staging

``` text
Package: dosbox-staging
Version: 0.75.1-2
Priority: optional
Section: otherosfs
Maintainer: Feignint <37488595+FeignInt@users.noreply.github.com>
Installed-Size: 3,257 kB
Depends: libasound2, libc6, libgcc-s1, libgl1, libopusfile0, libpng16-16, libsdl2-2.0-0, libsdl2-net-2.0-0, libstdc++6, zlib1g
Conflicts: dosbox
Homepage: https://dosbox-staging.github.io/about/
Distribution: unstable
Download-Size: 1,023 kB
APT-Sources: https://raw.githubusercontent.com/feignint/dosbox-staging-apt-repo-POC/master/generic unstable/otherosfs amd64 Packages
Description: x86 emulator for DOS-era video games
 dosbox-staging is an attempt to revitalize DOSBox's
 development process. It's not a rewrite, but a continuation
 and improvement on the existing DOSBox codebase while
 leveraging modern development tools and practices.
 .
 Yes I did copypasta that from the about page
```

### Install and run!

![Screenshot at 2020-06-07 20-36-37](https://user-images.githubusercontent.com/1557255/83990519-cea06880-a8fe-11ea-841e-2091d109075f.png)

**Comment by kcgen** [comment]
To get the `.deb` packages, is that a build and packaging series of steps you've manually run (or scripted) on your side?

Or is there a deb-creator script that digests the contents of `conf`, `db`, and `dists/unstable` and spits out the deb files?

**Comment by soredake** [comment]
```
curl https://raw.githubusercontent.com/feignint/dosbox-staging-apt-repo-POC/master/feignint.gpg | sudo apt-key add -
sudo add-apt-repository -s 'deb https://raw.githubusercontent.com/feignint/dosbox-staging-apt-repo-POC/master/generic unstable otherosfs'
```
And so far ppa and dosbox itself work fine.

**Comment by FeignInt** [comment]
the POC  is actually just reprepro
the only file I created in that was generic/conf/distributions

I then fed it the .changes file, and reprepro populated the rest
and I just cheapskate hosted it on github :D

I've not scripted the build, but it is certainly possible
Out-of-the-box, things like finalising a changelog entry spawns an editor,
if you look at 
```
apt changelog dosbox-staging
```
you should see that I link the release notes ( I got creative with  0.75.1 )
at the moment manual, but that is something I would want to automate

The actual build was,  using
[The debian/ tree found here](https://github.com/FeignInt/dosbox-staging-deb)

a rather rambling walkthrough
https://github.com/FeignInt/dosbox-staging-deb/wiki
when I was writing it I was mostly reminding myself , I've not been using git-buildpackage for long

```
gbp buildpackage -sa \
   --git-overlay \
   --git-pbuilder \
   --git-export-dir=../SomeBuildDir \
   --git-export=<debian-tag>
``` 
The debian tree lives on its own branch, in the same repo as dosbox-staging

gbp checksout  <debian-tag> 
it then translates the debian version found in the changelog to upstream tag.
tarballs the upstream tag, and the debian tag
the pbuilder bit builds in a chroot, which ensures a clean build

git buildpackage need not be used,
Simply dropping the debian/ tree into dosbox-staging and running
```
dpkg-buildpackage -b
```
is enough to build it,  
but I think gbp's overlay/export feature is cleaner

gbp has all kinds of hooks, i.e. a successful build can trigger a file push/upload etc.

Right now I'm happy that Mint was happy with a deb built in a DebianSid chroot

I'll poke around and see how the current builds are automated
just dropping the --git-pbuilder  is probably going to be fine assuming some flavour of debian is already building things

**Comment by FeignInt** [comment]
just a quick note on gbp workflow
The git-overlay may be over complicating things

something like this would be simpler

```
#!/bin/bash

TAG=v0.75.0-rc1
DEBTAG=${TAG:1}
DEBTAG=${DEBTAG//-/\~}

 PostEdit="sed -i '3{s@UNRELEASED@New upstream release:"
PostEdit+="\n  - https://dosbox-staging.github.io/"
PostEdit+="'${TAG//[.]/-}'/@}' debian/changelog"

gbp import-ref --upstream-tag=$TAG -u$DEBTAG || exit
gbp dch -N ${DEBTAG}-1 \
  -R \
  --spawn-editor=never \
  --postedit="$PostEdit" \
  --commit -- debian/ || exit
gbp tag || exit
gbp buildpackage -sa \
    --git-export-dir=../FooBuild \
    --git-export=debfoo/${DEBTAG/\~/_}-1 || exit
```
it is a rough script, just to demo automation
I hope it makes sense

**Comment by kcgen** [comment]
Thanks @FeignInt, this is exactly the help we need! I need to go through all your instructions in details, so I can ask informed questions.

But in general yes - as much headway as you can make on scripting this would be perfect. 

Here's the general context we can assume with the current CI chain:
1. We have a fresh clone of the dosbox-staging repo, and we're current sitting inside it's root directory (say: `/home/github-bot/dosbox-staging/`).
2. We have all tools and dependencies needed to compile dosbox-staging (so `./autogen.sh && ./configure && make` would run clean and produce a binary).
3. We have any of your static Debian-metadata config files in whatever directory tree they need, right along-side the source, including a GPG key.
4. We would then `apt-get install ... ` the debian specific tools needed to generate the `.deb` package, and any other outputs needed to satisfy the PPA/Debian world.
5. Next we would launch your new `Debian-package-and PPA-creator` script, that would do what you're outlining above.  Perhaps it would arrange everything in a subdirectory, say `./debian-ppa/*`
6. We would then figure out how to push the PPA tree to some github or gitlab area,  just like you've done.

The big help will be in 3 through 5; and you've probably already laid it all out; and I just need to read what you've done! 

My ideal lazy goal would be for you to attach a `tar.gz` and say: "extract this into your dosbox-staging sources, then run script `./scripts/deb-ppa-creator.sh`, and the PPA tree is ready to go inside `./debian/ppa` (or something like that.

**Comment by FeignInt** [comment]
ok, should not be too much of a problem
I've had a brief play with Ubuntu 18.04 via LXC

Things are a little different to Debian Sid, but no showstoppers.

I'm not certain how to deal with the gpg key
from the looks of things it is possible to add the private key as a "secret" on the repo., possibly having the passwd for the key as a second "secret". for the moment I will skip signing, and cross that bridge later.

I did stumble upon [persistent cache](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows)
It warns about storing sensitive information, since it could become publicly available , which is actually what we want
It probably isn't all that practical.

for actually getting both the new changelog, and the updated apt-repo to some usable state this looks promising
[actions create pull request ](https://github.com/marketplace/actions/create-pull-request)

I'm totally in the dark with all this CI Actions stuff
but the sketch in my head goes something like this

1. tag triggers CI action
2. script passed tag as argument
3. script generates new changelog, build && tags deb/version.
4. script adds to apt-repo
5. script opens PR for the "debian" branch and the  "apt-repo"
6. PRs merged as with any other PR

for now I will concentrate on {2..4} , trying to keep it as simple as possible while not being so simple it is prone to failing

**Comment by FeignInt** [comment]
[ this works on lxc bionic ](https://github.com/FeignInt/dosbox-staging-deb/tree/build_deb_pkgs/.github/scripts)

no logging or meaningful  error messages
and I have not gone out of my way to break it yet.
doesn't sign the apt repo bit, just tarballs them up

**Comment by kcgen** [comment]
Awesome progress!
 
I added the script to `.github/scripts/build-deb-pkgs.sh` on top of the master branch (with no other files from your repo), and launched it on my Ubuntu 20.04 box. 

``` bash
bash -x ./.github/scripts/build-deb-pkgs.sh v0.75.0
+ BaseDir=/home/kcroft
+ debian_branch=debianish
+ debian_worktree=/home/kcroft/debianish-worktree
+ reprepro_branch=reprepro
+ reprepro_base=/home/kcroft/reprepro-worktree
+ buildarea=/home/kcroft/buildarea/
+ debian_changelog=debian/changelog
+ [[ 1 != 1 ]]
+ case $1 in
+ case $0 in
+ [[ ! -d .git ]]
+ build_deb_pkgs=/usr/src/dosbox-staging/./.github/scripts/build-deb-pkgs.sh
+ dosbox_staging_git=/usr/src/dosbox-staging
+ upstreamTAG=v0.75.0
+ debVersion=v0.75.0-1
+ debVersion=0.75.0-1
+ debTAG=debianish/0.75.0-1
+ cd /usr/src/dosbox-staging
+ ListTag v0.75.0
++ git tag --list v0.75.0
+ [[ v0.75.0 == \v\0\.\7\5\.\0 ]]
+ ListTag debianish/0.75.0-1
++ git tag --list debianish/0.75.0-1
+ [[ '' == \d\e\b\i\a\n\i\s\h\/\0\.\7\5\.\0\-\1 ]]
+ init_worktree /home/kcroft/debianish-worktree debianish
+ grep -q '\[debianish\]'
++ git worktree list
+ git worktree add /home/kcroft/debianish-worktree debianish
fatal: invalid reference: debianish
+ exit 1
```

I know you're still working on it, so just providing on going feedback.  Tons of good stuff in the script.. will retest again when you want.

**Comment by kcgen** [comment]
Awesome progress!
 
I added the script to `.github/scripts/build-deb-pkgs.sh` on top of the master branch (with no other files from your repo), and launched it on my Ubuntu 20.04 box. 

``` bash
bash -x ./.github/scripts/build-deb-pkgs.sh v0.75.0
+ BaseDir=/home/kcgen
+ debian_branch=debianish
+ debian_worktree=/home/kcgen/debianish-worktree
+ reprepro_branch=reprepro
+ reprepro_base=/home/kcgen/reprepro-worktree
+ buildarea=/home/kcgen/buildarea/
+ debian_changelog=debian/changelog
+ [[ 1 != 1 ]]
+ case $1 in
+ case $0 in
+ [[ ! -d .git ]]
+ build_deb_pkgs=/usr/src/dosbox-staging/./.github/scripts/build-deb-pkgs.sh
+ dosbox_staging_git=/usr/src/dosbox-staging
+ upstreamTAG=v0.75.0
+ debVersion=v0.75.0-1
+ debVersion=0.75.0-1
+ debTAG=debianish/0.75.0-1
+ cd /usr/src/dosbox-staging
+ ListTag v0.75.0
++ git tag --list v0.75.0
+ [[ v0.75.0 == \v\0\.\7\5\.\0 ]]
+ ListTag debianish/0.75.0-1
++ git tag --list debianish/0.75.0-1
+ [[ '' == \d\e\b\i\a\n\i\s\h\/\0\.\7\5\.\0\-\1 ]]
+ init_worktree /home/kcgen/debianish-worktree debianish
+ grep -q '\[debianish\]'
++ git worktree list
+ git worktree add /home/kcgen/debianish-worktree debianish
fatal: invalid reference: debianish
+ exit 1
```

I know you're still working on it, so just providing on going feedback.  Tons of good stuff in the script.. will retest again when you want.

**Comment by FeignInt** [comment]
it is failing for you as you don't have the debianish branch

I'm not getting very far getting it to work on github
It gets as far as append_changelog
trouble is, I've not figured out how to get a logfile out of it, so I don't know why gbp dch fails.
I just mashed a yml together based on the "quickstart" one and the Upload tarball  from the linux.yml.

initially It failed setting up worktree, but a fetch seemed to resolve that

- gbp seems to be installed,  I don't get an exit from "gbp config dch"
- removing the postedit didn't fix it
- nor did not committing the changelog

I think I'll try pushing a pre-made changelog and see if the gbp buildpackage works, then start fresh armed with some experience.

but at least I can do away with the $0 fullpath detection bit I hacked in, I can locate things based on $GITHUB_WORKSPACE
then again, I might just do away with gbp caliing the script for postbuild hooks and just glob the changes file to reprepro

**Comment by FeignInt** [comment]
well, I think I got it to build

pushing a changelog, and turning off tagging seemed to do the trick
so I'm guessing it is readonly, I need to look at that "actions to PR" 

I'm now assuming I'm not getting logs due to script exit none zero, so it never gets to the artifact bit 

btw, if you do add the debianish branch, you will need to delete the 0.75.Xs from the  debian/changelog and remove the debianish/ tags

**Comment by kcgen** [comment]
1. Cloned dosbox-staging
1. Created a `debianish` branch
1. Copied your repo's `./debian` tree into `dosbox-staging/`, as part of the new `debianish` branch
1. Copied your `.github/scripts/build-deb-pkgs.sh` script also into the same path inside `dosbox-staging/`
1. Launched `./.github/scripts/build-deb-pkgs.sh v0.75.0`

    - `line 171: cd: ~/debianish-worktree: No such file or directory`, so I created this.  Suggest creating this (or using `mktemp -d` to guarantee a clean new directory). 

1. Relaunched

    - `line 52: gbp: command not found`, so I tried `sudo apt install gbp`, but package is not found. Looked in the script and found an `apt get` line, so installed all of thise.

 1. Relaunched

    - `grep: debian/changelog: No such file or directory`
    - `gbp:error: ~/debianish-worktree is not a git repository`

**Comment by FeignInt** [comment]
not sure why git-buildpackage didn't install initially
I did change  skip if installed test late on ( my original didn't fail if residual config present ) 
the new test is a bit spammy, I didn't send to /dev/null 

anyway

```
git clone https://github.com/dosbox-staging/dosbox-staging.git
cd dosbox-staging
git remote add \
  -f \
  -t debianish \
  -t build_deb_pkgs \
   feignint https://github.com/feignint/dosbox-staging-deb.git

git merge feignint/build_deb_pkgs --allow-unrelated-histories
```

tracks two branhces, debianish and build_deb_pkgs

I do have  builddeb and debianlike branches, ( which I will be deleting soon ) that I was using to learn about workflows
I did end up merging debianlike into builddeb, to rule out issue with trying to use a worktree, 
I digress...
 
since debianish already has changelogs  and debianish/TAGs for 0.75.0 etc. 
just make a new test tag

```
git tag v0.76.0-test
bash -x ./.github/scripts/build-deb-pkgs.sh v0.76.0-test
```
you can either add new tags, or cd to the debian-worktree and 
```
git reset HEAD~1
```
and delete tag(s)
```
git tag -d debianish/v0.76.0-test
# or
# while read tag;do git tag -d "$tag";done < <(git tag -l "v*76*-test" )
```

and yeah, I can use mktemp to setup the worktree dir
I should also cleanup using 
```
git worktree remove ${debian_worktree}
```

debian_worktree  works in the repo tree

it looks odd
```
dosbox-staging/
└─debian/
     └─debian/
         ├─changelog
         └─rules
```
but seems to work ( I'm not sure if it is git magic or gbp excluding ./debian/ when taring )

if you want to push the debs to a repo
the  ${reprepro_base} should point to the dir holding the git repo ( or a local web server like apache or  lighttpd )
and optionally add
```
SignWith: <gpg key id> 
```
to the  conf/distributions 
I'm not sure if script will wait for passwd to be entered, I haven't setup a gpg key on the LXC yet
then git add, commit , push

tip: 
I recently added CFLAGS to debian/rules

revert 63e27285 for reduced buildtime
you also need to add `DEBUILD_PREPEND_PATH="/usr/lib/ccache"` to `~/.devscripts` to get ccache working

**Comment by FeignInt** [comment]
finally got it built, tared and uploaded as artifact

silly pathnames due to my lazy tar -cf, but that was not my concern at the time.

now that I know about the GIT_WORKSPACE I can set the script to run differently
e.g. ,  when not on github default to --git-pbuilder, or strongly hint at it and be more careful about where buildarea is

**Comment by kcgen** [comment]
Awesome progress on this @FeignInt!  

Curious and excited to give it a go (no rush; whenever you think it's ready).  
Looking forward to learning about it, and then folding it into the dosbox-staging CI chain.

**Comment by FeignInt** [comment]
@kcgen I'm getting there .. slowly ...
seems there is some quirk with workflows, I spent too long trying to get a workflow "active", eventually I worked out that it needs push: branch: <branch>, which can then be removed 
at the moment I have it just looking at pushed tags.

The good news is, I've had actions push commits to the debian tree, and a branch on a different repo ( reprepro files that serve to apt )

1. tag triggers workflow
2. yml checksout dosbox-staging ( db-s ) and apt-repo
3. script creates/commits changelog on db-s debian-branch
4. script compiles, packages debs and updates apt-repo
5. script commits changes on db-s and apt-repo
6. yml uploads the 10 pkg files as artifacts # redundant really

As I arrived at that iteratively ( a lot of which didn't work ), I need to squash the commits, and rebase. I may just do a (diff | patch) ;git add -p

Then I need to get the signing side of things sorted out.

I'm probably going to put everything into the yml and retire the shell script.

**Comment by FeignInt** [comment]
I think I have sorted most of it out now

[  apt repo - holds the debs ](https://github.com/FeignInt/dosbox-staging-apt-repo-POC/tree/sandbox)
>Note: no longer "unstable otherosfs", it is now "bionic main", and the root of sandbox instead of generic

[ debian tree holds the debian packaging files](https://github.com/FeignInt/dosbox-staging-deb/tree/debianlike)
>Not the default branch "debianish"

[ copy of release/0.75.x, with scripts shoehorned in ](https://github.com/FeignInt/dosbox-staging-deb/tree/pkg_debs)

basically I deleted current .github/ , added mine and rebased the commit to just before v0.75.0-rc1. Then retagged, with original annotations ( made a feature of them  )

Half of my struggles was not understanding that the actions only use the .yml of the commit it got triggered by, so when I was pushing an old tag , stuff just didn't work.

[ "The Master" just the .github scripts](https://github.com/FeignInt/dosbox-staging-deb/tree/build_deb_pkgs)

I need to PLACEHOLDER the main vars
the vars are way out of sync with the testing branches above.

At the moment the yml is "hardcoded" with the branch names
I think I should set them in the header ( named: "set env"  at the moment) and have the body use the env. var.

in the shell script, debian_branch and reprepro_branch should be made to point to correct branches
but , it would make sense to have it get those from the yml, in the same way it gets TAG.

I also need to fix a point of failure, if I push to the apt-repo while it is building, it fails to push as HEAD is not what it expects.
So instead I need to checkout the apt repo just before it is ready to push. and have a few tries depending on the exit code git gives.

Signing, I managed to get it all working, including  password protected priv key. However, it is much easier to just use a passwordless key.

I need to put a "key due to expire" warning at the very end, the failed run email is not all that helpful ( would be nice if it at least gave the exit code ) but the bottom of the log would have a message telling you that the key was going to expire in X weeks

**Comment by FeignInt** [comment]
PPA build
https://launchpad.net/~feignint/+archive/ubuntu/db-s-poc

triggered by
https://github.com/FeignInt/dosbox-staging-deb/tree/pkg_debs/.github
and
https://github.com/FeignInt/dosbox-staging-deb/tree/debianlike

note: the .github commits on pkg_debs were rebased to just before rc1

I just need to remove or condition the apt-repo checkout as at the moment it needlessly checksout  a repo it never uses.

oh, and remove the gpg password caching nonsense , add sig expiry warning + other stuff

I also need to make lintian happy, it is complaining about the way I set LDFLAGS, the copyright file and the manpage

**Comment by dreamer** [comment]
@FeignInt Why the version is named 0.75.1-1? We haven't released 0.75.1 yet (but we will).

Why you needed to change the LDFLAGS? If the buildsystem is not appropriate for some reason, please let us know :)
Same for manpage - if something is missing/incorrect, we'very interested in adapting the fix asap.

I haven't looked in details into the package (I'm not a Debian person), but I'll take a peek given some time.

I hope the plan is to eventually change the repo name to ppa:feignint/dosbox-staging ?

**Comment by FeignInt** [comment]
@dreamer 

there was a commit on the release branch after the v0.75.0 tag, so I tagged it a point release

There is nothing wrong with the LDFLAGS
instead of the Debian's default `LDFLAGS=-Wl,-z,relro` I've set it `-Wl,--as-needed -Wall -pipe -fpch-preprocess` same as the current build.sh. lintian shows a warning because of the way I did it, it is just being fussy it frond on me explicitly setting it.

The only reason I tried arm build was I saw some buttons on PPA
I'll see how far I get using default compiler flags ( which should just work ). I don't know how much difference the compiler flags actually make real world, but I do know -msse4.2 breaks core2duo. 
I can then seek out correct flags for arm ( but can't test a build, unless I go qemu route )

The PPA account will be deleted it is just a proof of concept, @kcgen mentions above that both you and they have PPA accounts
but for branding probably best to have  ppa:dosbox-staging/{stable,alpha}

Since the PPA side of things turned out to be so straight forward. The most problems I had was registering the gpg key, turns out the best way is to paste ascii armoured pub into  https://keyserver.ubuntu.com/  instead of gpg --send-key as the guides suggest
I can greatly simplify the scripts by cutting out all the reprepro related stuff ( no more messing with access tokens and multiple git repos : ) )
I plan to re-write to a single yml script
I had never used github workflows or PPA before so I understand it better now.

The manpage error is trivial
>W: dosbox-staging: manpage-has-errors-from-man usr/share/man/man1/dosbox.1.gz  line 242: 'R' is a string (producing the registered sign), not a macro.

similar problem with line 112 

quick check , send std out to /dev/null
man docs/dosbox.1 >/dev/null

This appears to fix those errors, 
but I have never done manpage formatting before
```
--- a/docs/dosbox.1
+++ b/docs/dosbox.1
@@ -109,9 +109,9 @@ an Internal Program, a DOS command or an executable on a mounted drive.
 Output version information and exit. Useful for frontends.
 .TP
 .BI "\-\-editconf ["editor ]
-.R Open the default configuration file in a text editor. If no editor name
+.RB "Open the default configuration file in a text editor. If no editor name
 is given, then use the program from EDITOR environment variable,
-otherwise DOSBox will try to guess the name.
+otherwise DOSBox will try to guess the name."
 .TP
 .BI \-opencaptures " program"
 .RI "calls " program " with as  first parameter the location of the captures folder."
@@ -238,7 +238,7 @@ this as well!)
 .RS
 .TP
 .B \-All
-.R Reread directory structure for all drives.
+.RB "Reread directory structure for all drives."
 .TP
 .B Drive:
 .RB "Reread directory structure for drive " Drive:
```

**Comment by dreamer** [comment]
> @dreamer
> 
> there was a commit on the release branch after the v0.75.0 tag, so I tagged it a point release

Uhm, ok… but this might be somewhat misleading to the users… I presumed packagers will be more interested in tracking the tags and not branches… that's what tags are there for in fact. We'll see how it'll turn out when we'll be preparing for releasing 0.75.1.

> There is nothing wrong with the LDFLAGS
> instead of the Debian's default `LDFLAGS=-Wl,-z,relro` I've set it `-Wl,--as-needed -Wall -pipe -fpch-preprocess` same as the current build.sh. lintian shows a warning because of the way I did it, it is just being fussy it frond on me explicitly setting it.

You don't need `-Wall` for package builds (it's relevant for development, but not for packaging); but suit yourself. `-pipe` exists in our development build scripts to work around bugs on macOS; AFAIK there's no need for using it anywhere outside of macOS (especially since you're not counting the warnings summaries the way we do in our CI); I think you can skip this flag. `-fpch-preprocess` is turned on only if you have ccache (again, irrelevant for packaging).

I suggest sticking to the flags suggested by Debian - they are probably better suited for Debian users. The one big exception is `-O` compiler flag - we strongly recommend going with `-O3` (until we'll find a good way to generate and distribute runtime profiles to packagers).

> The only reason I tried arm build was I saw some buttons on PPA
> I'll see how far I get using default compiler flags ( which should just work ). I don't know how much difference the compiler flags actually make real world, but I do know -msse4.2 breaks core2duo.

They do make a small difference (I don't have exact numbers, but anecdotally just switching from -march=nehalem to -march-haswell gives me 1-3% speedier builds); but I wouldn't worry about this AT ALL right now. Go with Debian defaults - we would much rather have ARM-compatible Debian package than 1% speed increase. Once the package will be good and usable, then we can improve the build process to accommodate different architectures.

> The PPA account will be deleted it is just a proof of concept, @kcgen mentions above that both you and they have PPA accounts
> but for branding probably best to have
> 
> edit: new line, wrap looked ugly
> ppa:dosbox-staging/{stable,alpha}

I hope we'll get included in Debian and Ubuntu repos (the sooner the better) and there will be no long-term need for PPA. I would prefer to leave packaging to the distribution packagers - I trust them to make better decisions about their OS than me :) I will take care of my distro and create COPR package to speed up the inclusion in Fedora, but don't plan on maintaining PPA myself.

> The manpage error is trivial
> 
> > W: dosbox-staging: manpage-has-errors-from-man usr/share/man/man1/dosbox.1.gz  line 242: 'R' is a string (producing the registered sign), not a macro.
> 
> similar problem with line 112 (…)> 
> 
> This appears to fix those errors: (…)

I'll take a look, thanks! :)

**Comment by FeignInt** [comment]
yeah, I normally just leave defaults

with defaults arm64 built successfully 

https://launchpad.net/~feignint/+archive/ubuntu/db-s-poc2/+sourcepub/11382068/+listing-archive-extra

at the moment the arm64 is listed as "pending publication"
but can be downloaded from 
https://launchpad.net/~feignint/+archive/ubuntu/db-s-poc2/+build/19492930

I don't have anything to test it on

**Comment by soredake** [comment]
Any progress on this? 🥺

**Comment by FeignInt** [comment]
scheduled test run at
1. 16:00 utc - should be v0.75.0-rc1
2. 17:00 utc - should be v0.75.0
3. 18:00 utc - should "do nothing"  

It will push to
https://launchpad.net/~feignint/+archive/ubuntu/dosbox-staging-test1

it may take a while to build once it hits PPA, 

I have changed a few things since I last pushed to PPA, but I expect it will work

I ticked many boxes

- [x]  AMD x86-64 (amd64) 
- [x]  ARM ARMv8 (arm64)
- [x]  ARM ARMv7 Soft Float (armel)
- [x]  ARM ARMv7 Hard Float (armhf)
- [x] Intel x86 (i386)
- [x] PowerPC (powerpc)
- [x] PowerPC64 Little-Endian (ppc64el)
- [ ] RISC-V 64-bit (riscv64)
- [ ] IBM System z (s390x) 

based on past tests:

| Architecture                     | Builds             | Runs               | comment       |
|-                                 |:-:                 |:-:                 |-              |
|AMD x86-64 (amd64)                | :heavy_check_mark: | :heavy_check_mark: | bullseye/sid, |
|Intel x86 (i386)                  | :heavy_check_mark: | :grey_question:    |               |
|ARM ARMv8 (arm64)                 | :heavy_check_mark: | :grey_question:    |               |
|ARM ARMv7 Soft Float (armel)      | :grey_question:    | :grey_question:    |               |
|ARM ARMv7 Hard Float (armhf)      | :grey_question:    | :grey_question:    |               |
|PowerPC (powerpc)                 | :grey_question:    | :grey_question:    |               |
|PowerPC64 Little-Endian (ppc64el) | :grey_question:    | :grey_question:    |               |

**Comment by dreamer** [comment]
@FeignInt forget about v0.75.0-rc1 - nobody cares about this old RC now once v0.75.0 is released :)

I see you skipped s390x - it's ok, I doubt there are any users interested in running dosbox-staging on this architecture; we do include it in our platform builds though, just to have test build process on at least one big-endian arch on regular basis ;)

Also, I think we'll do 0.75.1 release soon (as in a week or two - we'll see) - it includes important bugfixes (some commits were backported already to `release/0.75.x` branch, but we haven't tagged it yet).

**Comment by FeignInt** [comment]
regards 0.75.0-rc1,  yeah, I can just pre-tag so it is skipped

The un-ticked Archs on PPA are ones which I couldn't tick, I think you have to ask an admin to "unlock" them, but like you said I doubt anyone  would want/need them.

unfortunately it does seem to have failed on the PPA side
It uploaded but launchpad appears to have ignored it, I'm guessing it is an issue with the gpg signature.

I'll hopefully get a chance to figure out why later today

I will turn off the next run until I figure it out.

**Comment by FeignInt** [comment]
Figured it out in the end
I neglected to import a hunk which renamed PPA_ID to PPA_REPO in push_to_ppa()

https://launchpad.net/~feignint/+archive/ubuntu/dosbox-staging-test1/+sourcepub/11409782/+listing-archive-extra

-   [x] amd64 - Pending publication
-   [x] arm64 - Pending publication
-   [x] armhf - Pending publication
-   [x] i386 - Pending publication
-   [x] ppc64el - Pending publication

I'm not sure why armel and powerpc are missing, I don't see errors for them.
but al least the amd64 and arm64 seem to be ok.

I'll leave that PPA up for a while, until I sort out a cleaner one, with full descriptions/links etc.

**Comment by FeignInt** [comment]
https://launchpad.net/~feignint/+archive/ubuntu/dosbox-staging

https://github.com/FeignInt/dosbox-staging-deb/tree/debianish

*almost* fully automated.

At the moment if won't account for changes to build-depends, maybe I can script that, I'm not sure.

**Comment by soredake** [comment]
@FeignInt only builds for bionic? No focal?

**Comment by FeignInt** [comment]
is focal 20.04 ?

at the moment it uses ubuntu-latest, which is 18.04
I guess I could do one of those marix things to do ubuntu-20.04 , and any others
  
The codename is not all that important, I stripped reference to minimum version number, it should work.
You could compile/package yourself, some details on that included in the READ.md on the debianish branch

**Comment by FeignInt** [comment]
update:

- [ dosbox-staging release PPA ](https://launchpad.net/~feignint/+archive/ubuntu/dosbox-staging)

  - Groovy, Focal, Eoan, Bionic, Xenial, Trusty
  
- [ dosbox-staging-weekly PPA ](https://launchpad.net/~feignint/+archive/ubuntu/dosbox-staging-weekly)
  - Groovy, Focal 

The weekly is only amd64, I can tick more arches if needed. I can also have it package older Ubuntu series. At the moment it is a separate .yml workflow, but I expect I will merge the two workflows into one file with a re-write.

I have added support for fluidsynth, but this will only be enabled on Focal and newer as Fluidsynth version is < 2.0.0 pre Focal, as far as I can see no v2 backports.

---
Note: I have not *added* fluidsynth as in a *cherrypick*, just included build depends and a conditional to add *--disable-fluidsynth* if version is too low.

---
The first weekly changelog is huge ( all commits since v0.75.0 ), the next will only list changes since the last weekly. I do need still need to filter the changelog entries, for instance to ignore changes in *.github, scripts etc.* not certain I can do that easily,

Currently the weekly and release can not be installed side-by-side.

In the future  I could adjust things so they are separate independent packages e.g. _--program-suffix=-alpha_ and have _debian/rules_ use control.in templates to amend package-name, description etc.

**Comment by kcgen** [comment]
Really slick @FeignInt! 

I like the ascii art; that's definitely another lost gem from the early 90s BBS, release, and demo scenes.

 Regarding platforms, I think amd64 will cover the 99%, and if users have a way to request something else, we could gauge interest that way.  Something like "_Don't see your platform or OS version? Contact us at xyz_" 

The weekly releases are perfect for users who want to keep on the latest code but don't want the overhead of doing their own builds; just let the auto updater keep you in sync. I also agree that it's probably fine for this to conflict with the stable release. 

 If someone really wants to flip between the two versions, then they'll probably be building from source anyway and even checking out specific commits.

That's a nice changelog summary; is that something you've added or is it formatted by part the ppa tooling?

https://launchpadlibrarian.net/489902812/dosbox-staging_0.76.0~alpha~332~g6694-1+u20.10_source.changes

**Comment by FeignInt** [comment]
Yeah, I though the ascii was fun

I did try colour ascii , it worked but made the ppa page unreadable as *asciiart* set the color code for each char.,  I have since manually tamed it a bit ( only change colorcode when needed) maybe I will  add that later.

The changelog is from git-buildpackage ( gbp dch --since=<commit/tag>, but I had to hack it a little as by default it will only consider the branch you are on.

The hacks are in `hack_dch_py()` and `patched_dch()` in *weekly_PPA.yml* 
Currently sed. I need to dig deeper and possibly produce a cleaner more complete patch(set) for upstream

`gbp dch --verbose` shows a little of what is going on under the hood, basically `git rev-parse` and `git show --pretty=format:...`
But it also has some handy metatag parsing for including Fixes:  and Thanks:
